tests/Http/BookTest.php

```diff
@@ -94,4 +94,20 @@ public function testTitlePrefill() {
 		$this->assertStringContainsString( '<input name="page" id="page" type="text" size="30" required="required" class="form-control"
 					value="A &quot;title&quot;" />', $client->getResponse()->getContent() );
 	}
+
+	public function testFonts() {
+		$client = static::createClient();
+		// No font.
+		$client->request( 'GET', '/', [ 'fonts' => 'Not a font name' ] );
+		$this->assertStringContainsString( '<option value="" selected="selected">None (use device default)</option>', $client->getResponse()->getContent() );
+		// Default when there's no on-wiki config.
+		$client->request( 'GET', '/', [ 'lang' => 'en' ] );
```

Comment:
Isn't this gonna fail when the default font is set on enwikisource?

Reply:
No, the idea is that if you give it an invalid font name it *doesn't* fall back to the default, it just doesn't embed one. Should it use the default in that situation? Also, I think it should be possible to choose *not* to embed a font (which hasn't been the case for non-latin languages); that's a separate issue I think.

Reply:
What I mean is that on line 103 it will go and fetch the defaultFont on en.wikisource.org. If I create that json file right now, the assert on 104 will fail because it will actually find a default font. Does that makes sense? Just like with beta, if I change the defaultFont there line 106 will start failing. Am I missing something?

Everything else looks great btw

Reply:
That's a good point.

I sort of think this isn't too much of an issue because a) it's unlikely that English Wikisource will set a default font; and b) this whole test class is based on on-wiki data that might change.

The trouble here is that to mock this isn't too straight forward, because this is meant to be the highest level of integration tests (that test the whole stack). I've tried mocking and changing the test client's container, but it doesn't work:

```php
		$client = static::createClient();
		$onWikiConfig = $this->createMock( OnWikiConfig::class );
		$onWikiConfig->expects( $this->once() )
			->method( 'getDefaultFont' )
			->willReturn( 'foo' );
		$client->getContainer()->set( OnWikiConfig::class, $onWikiConfig );
```

From what I read, this is generally considered a bad idea.

Any ideas?

Reply:
I just tried it locally and I was able to run the test with some minor changes. [Here is what I tried](https://github.com/wikimedia/ws-export/files/6172784/patch.txt)

Why is it considered a bad idea? Typically, you wouldn't mock services when running integration tests but external resources that we can't control are an exception IMO unless you can guarantee that the external config is not gonna change. @cimurah, @MusikAnimal, any thoughts?

For an even more "stricter" test you could mock API response instead of OnWikiConfig

In any case, the rest of the patch is good to go and I'm happy to merge it if you want to keep it as it is.

Reply:
I really don't have much to say since tests can still be confusing for me.  I need more experience around mocking.

> Why is it considered a bad idea? Typically, you wouldn't mock services when running integration tests but external resources that we can't control are an exception IMO unless you can guarantee that the external config is not gonna change. @cimurah, @MusikAnimal, any thoughts?

I do want to add that reading this review alone helped a lot. I understand what @dayllanmaza meant with the example he provided, but not sure if it's necessary for this case.

Reply:
@dayllanmaza That's pretty much what I was trying too, but it doesn't work. Try changing your code to `->will( $this->onConsecutiveCalls( 'foo', 'bar', 'baz', 'bif' ) );` and you'll see the tests still pass — they're still using the actual OnWikiConfig service and making the call to the API.

As for why it's a bad idea, I'm sort of on the fence about it, but I think the idea is that functional tests should be thought of as high-level integration tests, and that it's more about making sure everything works together and treading the system itself as a black box. Anyway, I'd be happy to do it, but can't figure out how to get it working.